File: src/data/shopping_cart_speedrun_100_wr.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

int main() {
    int N, M;
    cin >> N >> M;

    vector<int> pods;
    for (int i = 0; i < M; ++i) {
        int X;
        cin >> X;
        pods.push_back(X);
    }

    int ans = 2147483647, ct = 0;
    unordered_map<int, int> cart;
    for (int l = 0, r = 0; r < M; ++r) {
        ct += !(cart[pods[r]]++);
        while (ct == N) {
            ans = min(ans, r-l+1);
            ct -= !(--cart[pods[l++]]);
        }
    }

    cout << ans << "\n";
}

File: src/data/shopping_cart_speedrun_100_wr.py
N, M = map(int, input().split())
pods = list(map(int, input().split()))

ans, count = 10**9, 0
cart = [0] * N

l = 0
for r in range(M):
    if cart[pods[r]] == 0:
        count += 1;
    cart[pods[r]] += 1
    while count == N:
        ans = min(ans, r-l+1);
        cart[pods[l]] -= 1
        if cart[pods[l]] == 0:
            count -= 1
        l += 1

print(ans)
